node_modules/gh-auth/lib/oauth.js

```diff
@@ -0,0 +1,516 @@
+                // we will format the URI afterward
+                parsedRedirectUri.query = _.chain(parsedRedirectUri.query)
+                    .ghOptional()
+                    .extend({'code': code})
+                    .value();
+                parsedRedirectUri.search = null;
+
+                log().info({
+                    'actorId': ctx.user.id,
+                    'clientId': clientId,
+                    'forUserId': userId,
+                    'expires': expires
+                }, 'OAuth Authorization Code has been granted for a user');
+
+                return callback(null, url.format(parsedRedirectUri));
+            });
+        });
+    });
+};
+
+/**
+ * Exchange an authorization code for an access token. In the client access token exchange, a client
+ * secret is provided. Since the client id and secret are the "authentication" mechanisms (i.e.,
+ * username and password), the check that applies to the secret should have already been done before
+ * this function gets invoked
+ *
+ * @param  {OauthClient}    client          The *authenticated* oauth client that is performing the request
+ * @param  {String}         code            The code to exchange for an access token
+ * @param  {String}         redirectUri     The redirect uri of the client
+ * @param  {Function}       callback        Standard callback function
+ * @param  {Object}         callback.err    An error that occurred, if any
+ * @param  {String}         callback.token  The access token that can be used to perform "bearer" authenticated requests
+ */
+var exchangeAuthCode = module.exports.exchangeAuthCode = function(client, code, redirectUri, callback) {
+    var validator = new Validator();
+    validator.check(code, {'code': 400, 'msg': 'An authorization code must be specified'}).notEmpty();
+    validator.check(client, {'code': 400, 'msg': 'A valid client must be specified'}).notEmpty();
+    validator.check(redirectUri, {'code': 400, 'msg': 'A valid redirect URI must be specified'}).isUrl();
+    if (validator.hasErrors()) {
+        return callback(validator.getFirstError());
+    } else if (client.disabled) {
+        return callback({'code': 401, 'msg': 'The specified client has been disabled'});
+    } else if (client.redirectUri !== redirectUri) {
+        return callback({'code': 401, 'msg': util.format('The specified redirect uri "%s" does not match the redirect uri configured for the client')});
+    }
+
+    AuthOauthDAO.getAuthCodeByCode(code, function(err, authCode) {
+        if (err) {
+            return callback(err);
+        } else if (authCode.OauthClientId !== client.id) {
+            return callback({'code': 401, 'msg': 'The provided authorization code does not map to the specified client'});
+        } else if (authCode.expires.getTime() <= Date.now()) {
+            // TODO, this shouldn't even work? (authCode.expires.getTime())
```

Comment:
I think there was a time when it looked like those timestamps weren't going to be `Date`'s